### PR TITLE
sync blobs in initial-sync

### PR DESCRIPTION
**Pull request by kasey**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

Adds the ability for initial-sync to work past deneb by syncing blobs for ranges of blocks that require them.

### CODE REVIEW

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 1]
```diff
@@ -293,17 +309,93 @@ func (f *blocksFetcher) fetchBlocksFromPeer(
...
+			continue
 		}
+		f.p2p.Peers().Scorers().BlockProviderScorer().Touch(p)
+		return blocks, p, err
 	}
 	return nil, "", errNoPeersAvailable
 }
 
+func blobRequestsForBlocks(blocks []interfaces.ReadOnlySignedBeaconBlock) (*p2ppb.BlobSidecarsByRangeRequest, error) {
+	log.WithField("block-len", len(blocks)).Warn("blobRequestsForBlocks 1")
```
Comment (prestonvanloon) [inline beacon-chain/sync/initial-sync/blocks_fetcher.go:325]: Is this a meaningful log message?

File: beacon-chain/sync/initial-sync/round_robin.go [hunk 2]
```diff
@@ -244,48 +266,96 @@ func (s *Service) processBlock(
...
+					root, ci, kc, blobi.KzgCommitment)
+			}
+		}
+	}
+
+	if len(unprocessed) == 0 {
+		maxIncoming := blks[len(blks)-1]
+		maxRoot, err := maxIncoming.Block().HashTreeRoot()
+		if err != nil {
+			log.Errorf("TODO: use ROBlock so this kind of silly error handling isn't needed")
```
Comment (prestonvanloon) [inline beacon-chain/sync/initial-sync/round_robin.go:338]: Want to handle this TODO?

File: proto/engine/v1/json_marshal_unmarshal.go [hunk 3]
```diff
@@ -108,10 +111,32 @@ func (e *ExecutionBlock) UnmarshalJSON(enc []byte) error {
...
+			}
+			err = u.UnmarshalText([]byte(sedg))
+			if err != nil {
+				return errors.Wrap(err, "unable to unmarshal excessDataGas as hexutil.Uint64")
+			}
+			e.ExcessDataGas = u
 		}
+
 		dgu, has := decoded["dataGasUsed"]
+		log.Error(has, dgu != nil)
```
Comment (prestonvanloon) [inline proto/engine/v1/json_marshal_unmarshal.go:127]: Looks like a debug log. Want to amend this?

File: beacon-chain/sync/subscriber.go [hunk 4]
```diff
@@ -201,8 +201,10 @@ func (s *Service) subscribeWithBase(topic string, validator wrappedVal, handle s
 		defer func() {
 			if r := recover(); r != nil {
 				tracing.AnnotateError(span, fmt.Errorf("panic occurred: %v", r))
-				log.WithField("error", r).Error("Panic occurred")
-				debug.PrintStack()
+				log.WithField("error", r).
+					WithField("recovered_at", "subscribeWithBase").
+					WithField("stack", string(debug.Stack())).
```
Comment (prestonvanloon) [inline beacon-chain/sync/subscriber.go:206]: nice improvement for log ingestion

File: beacon-chain/sync/initial-sync/round_robin.go [hunk 5]
```diff
@@ -244,48 +262,67 @@ func (s *Service) processBlock(
 	return blockReceiver(ctx, blk, blkRoot)
 }
 
+var ErrMissingBlobsForBlockCommitments = errors.New("blobs unavailable for processing block with kzg commitments")
+var ErrBlobCommitmentMismatch = errors.New("kzg_commitment in blob does not match block")
+var ErrUnexpectedBlobCommitment = errors.New("kzg_commitment in blob not referenced in block")
+
+type processedChecker func(context.Context, blocks.ROBlock) bool
+
+func validUnprocessed(ctx context.Context, bwb []BlockWithVerifiedBlobs, headSlot primitives.Slot, isProc processedChecker) ([]BlockWithVerifiedBlobs, error) {
```
Comment (rauljordan) [inline beacon-chain/sync/initial-sync/round_robin.go:271]: This pure function could benefit from a unit test that covers different edge cases

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 6]
```diff
@@ -293,17 +312,159 @@ func (f *blocksFetcher) fetchBlocksFromPeer(
...
+	// Loop through the blocks, looking for the earliest deneb block in the slice.
+	for i, b := range bwb {
+		slot := b.block.Block().Slot()
+		if slot < denebStart {
+			continue
+		}
+		commits, err := b.block.Block().Body().BlobKzgCommitments()
+		if err != nil {
+			if errors.Is(err, consensus_types.ErrUnsupportedField) {
+				continue
```
[the remark this answers is not in the document]
Comment (kasey) [reply beacon-chain/sync/initial-sync/blocks_fetcher.go:433]: yes, this shouldn't happen. an unfortunate consequence of having to handle the errors that can come from these block-superset getters. I added a warning log for this case.

File: beacon-chain/blockchain/process_block.go [hunk 7]
```diff
@@ -144,19 +144,14 @@ func getStateVersionAndPayload(st state.BeaconState) (int, interfaces.ExecutionD
 	return preStateVersion, preStateHeader, nil
 }
 
-func (s *Service) onBlockBatch(ctx context.Context, blks []interfaces.ReadOnlySignedBeaconBlock,
-	blockRoots [][32]byte) error {
+func (s *Service) onBlockBatch(ctx context.Context, blks []consensusblocks.ROBlock) error {
```
Comment (potuz) [inline beacon-chain/blockchain/process_block.go:147]: this is not related to this PR, but I would prefer the type to signal that the block is signed.

File: beacon-chain/blockchain/process_block.go [hunk 8]
```diff
@@ -255,13 +251,13 @@ func (s *Service) onBlockBatch(ctx context.Context, blks []interfaces.ReadOnlySi
 			JustifiedCheckpoint: jCheckpoints[i],
 			FinalizedCheckpoint: fCheckpoints[i]}
 		pendingNodes[len(blks)-i-1] = args
-		if err := s.saveInitSyncBlock(ctx, blockRoots[i], b); err != nil {
+		if err := s.saveInitSyncBlock(ctx, root, b); err != nil {
```
Comment (potuz) [inline beacon-chain/blockchain/process_block.go:254]: This function can now be modified to not take the root and use b itself.

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 9]
```diff
@@ -293,17 +313,195 @@ func (f *blocksFetcher) fetchBlocksFromPeer(
...
+		rb, err := blocks2.NewROBlock(b)
+		if err != nil {
+			return nil, err
+		}
+		robs[i] = rb
+	}
+	sort.Sort(blocks2.ROBlockSlice(robs))
+	rb := make([]BlockWithVerifiedBlobs, len(robs))
+	for i := range robs {
+		rb[i] = BlockWithVerifiedBlobs{block: robs[i]}
```
Comment (potuz) [inline beacon-chain/sync/initial-sync/blocks_fetcher.go:345]: It's a little confusing to me why we need to make the slice `robs`  to then sort it to then copy the contents into the slice `rb`. Couldn't we simply implement the sort interface in `[]BlockWithVerifiedBlobs` and then create directly the slice of verified blocks?

File: beacon-chain/blockchain/process_block.go (hunk 7 above)
Comment (terencechain) [reply beacon-chain/blockchain/process_block.go:147]: maybe this can be another PR against `develop` if we decide to go with RO block route here

File: beacon-chain/blockchain/testing/mock.go [hunk 10]
```diff
@@ -182,7 +183,7 @@ func (s *ChainService) ReceiveBlockInitialSync(ctx context.Context, block interf
 	}
 	parentRoot := block.Block().ParentRoot()
 	if !bytes.Equal(s.Root, parentRoot[:]) {
-		return errors.Errorf("wanted %#x but got %#x", s.Root, block.Block().ParentRoot())
+		return errors.Errorf("ReceiveBlockInitialSync wanted %#x but got %#x", s.Root, block.Block().ParentRoot())
```
Comment (terencechain) [inline beacon-chain/blockchain/testing/mock.go:186]: is this an intended change or just for debug?

File: beacon-chain/execution/engine_client.go [hunk 11]
```diff
@@ -714,7 +723,7 @@ func fullPayloadFromExecutionBlock(
 		txs[i] = txBin
 	}
 
-	switch block.Version {
+	switch blockVersion {
```
Comment (terencechain) [inline beacon-chain/execution/engine_client.go:726]: why doesn't block.version work?

Comment (kasey) [reply beacon-chain/execution/engine_client.go:726]: The intent of this change was to replace code that relied on heuristics inside the execution block unmarshal function to guess the version of the block, because this code was buggy, which led to the original type to be mutated. Since the caller of this function knows the original blinded type, I thought it made more sense to let the caller specify the type to use, and avoid the buggy guessing game entirely.

File: beacon-chain/blockchain/testing/mock.go (hunk 10 above)
Comment (kasey) [reply beacon-chain/blockchain/testing/mock.go:186]: I think we had a few "wanted x but got y" error messages that made it difficult to suss out the source of the error. I can see not liking the func name in the error message but it's nice to have errors that can be traced back to their source. I don't feel strongly on whether to revert this change or leave it in.

Comment (terencechain) [reply beacon-chain/blockchain/testing/mock.go:186]: I don't know what is idiomatic here, but I haven't seen much of the copying/pasting function as it is. I thought the common pattern would be more descriptive `failed to compare root in receive block initial sync`... etc but I don't feel strongly either

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 12]
```diff
@@ -334,6 +532,30 @@ func (f *blocksFetcher) requestBlocks(
...
+	l := f.peerLock(pid)
+	l.Lock()
+	log.WithFields(logrus.Fields{
+		"peer":     pid,
+		"start":    req.StartSlot,
+		"count":    req.Count,
+		"capacity": f.rateLimiter.Remaining(pid.String()),
+		"score":    f.p2p.Peers().Scorers().BlockProviderScorer().FormatScorePretty(pid),
+	}).Debug("Requesting blobs")
+	if f.rateLimiter.Remaining(pid.String()) < int64(req.Count) {
```
Comment (nisdas) [inline beacon-chain/sync/initial-sync/blocks_fetcher.go:548]: The rate limiter here is only for blocks, do we want to use it for blobs too ?

File: beacon-chain/sync/initial-sync/blocks_fetcher_utils.go [hunk 13]
```diff
@@ -226,37 +233,51 @@ func (f *blocksFetcher) findForkWithPeer(ctx context.Context, pid peer.ID, slot
...
 	}
 	blocks, err := f.requestBlocks(ctx, req, pid)
 	if err != nil {
 		return nil, fmt.Errorf("cannot fetch blocks: %w", err)
 	}
+	if len(blocks) == 0 {
+		return nil, errNoAlternateBlocks
+	}
+	first := blocks[0]
+	if !f.chain.HasBlock(ctx, first.Block().ParentRoot()) {
```
Comment (nisdas) [inline beacon-chain/sync/initial-sync/blocks_fetcher_utils.go:247]: Why do we have this here ?

File: beacon-chain/sync/initial-sync/blocks_fetcher_utils.go [hunk 14]
```diff
@@ -226,37 +233,51 @@ func (f *blocksFetcher) findForkWithPeer(ctx context.Context, pid peer.ID, slot
...
-			return &forkData{peer: pid, blocks: blocks}, nil
+		if f.chain.HasBlock(ctx, parentRoot) {
+			continue
+		}
+		log.WithFields(logrus.Fields{
+			"peer": pid,
+			"slot": block.Block().Slot(),
+			"root": fmt.Sprintf("%#x", parentRoot),
+		}).Debug("Block with unknown parent root has been found")
+		altBlocks, err := validROBlocks(blocks[i-1:])
```
Comment (nisdas) [inline beacon-chain/sync/initial-sync/blocks_fetcher_utils.go:268]: Don't understand this, why will we want to fetch blobs for these blocks ? Don't they have unknown parents

File: beacon-chain/sync/initial-sync/round_robin.go [hunk 15]
```diff
@@ -244,48 +262,63 @@ func (s *Service) processBlock(
...
+		maxRoot := maxIncoming.Root()
+		return nil, fmt.Errorf("headSlot:%d, blockSlot:%d , root %#x:%w", headSlot, maxIncoming.Block().Slot(), maxRoot, errBlockAlreadyProcessed)
+	}
+	return bwb[*processed:], nil
+}
+
 func (s *Service) processBatchedBlocks(ctx context.Context, genesis time.Time,
-	blks []interfaces.ReadOnlySignedBeaconBlock, bFunc batchBlockReceiverFn) error {
-	if len(blks) == 0 {
+	bwb []BlockWithVerifiedBlobs, bFunc batchBlockReceiverFn) error {
```
Comment (nisdas) [inline beacon-chain/sync/initial-sync/round_robin.go:296]: a lot of these changes might be better suited to be directly against develop, as they seem to be bigger refactors rather than something deneb specific

File: beacon-chain/sync/initial-sync/round_robin.go [hunk 16]
```diff
@@ -244,48 +262,63 @@ func (s *Service) processBlock(
...
-	s.logBatchSyncStatus(genesis, blks, blkRoot)
-	parentRoot := firstBlock.Block().ParentRoot()
-	if !s.cfg.Chain.HasBlock(ctx, parentRoot) {
-		return fmt.Errorf("%w: %#x (in processBatchedBlocks, slot=%d)", errParentDoesNotExist, firstBlock.Block().ParentRoot(), firstBlock.Block().Slot())
-	}
-	blockRoots := make([][32]byte, len(blks))
-	blockRoots[0] = blkRoot
-	for i := 1; i < len(blks); i++ {
-		b := blks[i]
-		if b.Block().ParentRoot() != blockRoots[i-1] {
```
Comment (nisdas) [inline beacon-chain/sync/initial-sync/round_robin.go:278]: this change is being removed, is this asserted anywhere ?

File: beacon-chain/blockchain/process_block.go (hunk 8 above)
Comment (nisdas) [reply beacon-chain/blockchain/process_block.go:254]: I guess more on changing the block type from `SignedBeaconBlock` to `RoBlock` across initial sync

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 17]
```diff
@@ -293,17 +313,195 @@ func (f *blocksFetcher) fetchBlocksFromPeer(
...
+	return blobs
+}
+
+var errBlobVerification = errors.New("peer unable to serve aligned BlobSidecarsByRange and BeaconBlockSidecarsByRange responses")
+var errMissingBlobsForBlockCommitments = errors.Wrap(errBlobVerification, "blobs unavailable for processing block with kzg commitments")
+var errMismatchedBlobBlockRoot = errors.Wrap(errBlobVerification, "BlockRoot in BlobSidecar does not match the expected root")
+var errMissingBlobIndex = errors.Wrap(errBlobVerification, "missing expected blob index")
+var errMismatchedBlobCommitments = errors.Wrap(errBlobVerification, "commitments at given slot, root and index do not match")
+
+func verifyAndPopulateBlobs(bwb []BlockWithVerifiedBlobs, blobs []*p2ppb.BlobSidecar, blobWindowStart primitives.Slot) ([]BlockWithVerifiedBlobs, error) {
```
[the remark this answers is not in the document]
Comment (kasey) [reply beacon-chain/sync/initial-sync/blocks_fetcher.go:414]: I don't have benchmarks for this. What part seems expensive? We ultimately need to check all the blobs against the block commitments, and while doing so I thought it made sense to pair them so further processing could assume the validation is complete. Does something about the implementation seem naively inefficient, like there is a way to accomplish what this function does in a more efficient manner?

File: beacon-chain/sync/initial-sync/blocks_fetcher.go (hunk 12 above)
Comment (kasey) [reply beacon-chain/sync/initial-sync/blocks_fetcher.go:548]: My thinking was that since we want to request blocks and blobs from the same peer, we should combine these limits, since the requests stack on each other.

File: beacon-chain/sync/initial-sync/blocks_fetcher_utils.go (hunk 14 above)
Comment (kasey) [reply beacon-chain/sync/initial-sync/blocks_fetcher_utils.go:268]: My understanding of this code path is that when we detect get stuck syncing blocks, we look for peers with a different head and then try to backtrack to a common ancestor so that we can attempt to sync the alternate branch. In this change I am just trying to clean up the existing implementation a little because the loop was written in a super weird way and add blob support. Maybe we could walk through this code more carefully to review together in a video call.

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 18]
```diff
@@ -293,17 +313,195 @@ func (f *blocksFetcher) fetchBlocksFromPeer(
...
+		bb.blobs = make([]*p2ppb.BlobSidecar, len(commits))
+		for ci := range commits {
+			// There are more expected commitments in this block, but we've run out of blobs from the response
+			// (out-of-bound error guard).
+			if blobi == len(blobs) {
+				return nil, missingCommitError(bb.block.Root(), commits[ci:])
+			}
+			bl := blobs[blobi]
+			if bl.Slot != block.Slot() {
+				return nil, missingCommitError(bb.block.Root(), commits[ci:])
```
Comment (rkapka) [inline beacon-chain/sync/initial-sync/blocks_fetcher.go:446]: Is this the correct error? Shouldn't we say `block slot %d != BlobSidecar.Slot %d at slot %d`?

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 19]
```diff
@@ -293,17 +313,195 @@ func (f *blocksFetcher) fetchBlocksFromPeer(
...
+					Warn("block with slot within blob retention period has version which does not support commitments")
+				continue
+			}
+			return nil, err
+		}
+		bb.blobs = make([]*p2ppb.BlobSidecar, len(commits))
+		for ci := range commits {
+			// There are more expected commitments in this block, but we've run out of blobs from the response
+			// (out-of-bound error guard).
+			if blobi == len(blobs) {
```
Comment (rkapka) [inline beacon-chain/sync/initial-sync/blocks_fetcher.go:441]: What if we get more blobs than there are commitments in the block? Assuming that's even possible.

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 20]
```diff
@@ -293,17 +313,195 @@ func (f *blocksFetcher) fetchBlocksFromPeer(
...
+	for k := range missing {
+		missStr = append(missStr, fmt.Sprintf("%#x", k))
+	}
+	return errors.Wrapf(errMissingBlobsForBlockCommitments,
+		"block root %#x missing %d commitments %s", root, len(missing), strings.Join(missStr, ","))
+}
+
+// fetchBlobsFromPeer fetches blocks from a single randomly selected peer.
+func (f *blocksFetcher) fetchBlobsFromPeer(ctx context.Context, bwb []BlockWithVerifiedBlobs, pid peer.ID) ([]BlockWithVerifiedBlobs, error) {
+	ctx, span := trace.StartSpan(ctx, "initialsync.fetchBlobsForResponse")
```
Comment (rkapka) [inline beacon-chain/sync/initial-sync/blocks_fetcher.go:482]:
```suggestion
	ctx, span := trace.StartSpan(ctx, "initialsync.fetchBlobsFromPeer")
```

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 21]
```diff
@@ -293,17 +313,195 @@ func (f *blocksFetcher) fetchBlocksFromPeer(
...
+	}
+	// Construct request message based on observed interval of blocks in need of blobs.
+	req := blobRequest(bwb, blobWindowStart)
+	if req == nil {
+		return bwb, nil
+	}
+	// Request blobs from the same peer that gave us the blob batch.
+	blobs, err := f.requestBlobs(ctx, req, pid)
+	if err != nil {
+		return nil, errors.Wrap(err, "Could not request blobs by range")
```
Comment (rkapka) [inline beacon-chain/sync/initial-sync/blocks_fetcher.go:499]:
```suggestion
		return nil, errors.Wrap(err, "could not request blobs by range")
```

File: beacon-chain/sync/initial-sync/blocks_fetcher_utils.go (hunk 14 above)
Comment (kasey) [reply beacon-chain/sync/initial-sync/blocks_fetcher_utils.go:268]: I'm going to add comments to this method to clarify why we're doing this.

File: beacon-chain/sync/initial-sync/blocks_fetcher.go (hunk 12 above)
Comment (kasey) [reply beacon-chain/sync/initial-sync/blocks_fetcher.go:548]: Note from conversation with Nishant: we should do some napkin math on the relative block vs blob sizes so that we can apply the right rate limit here. We have tentative agreement on the idea that overriding the block rate limit for blobs accomplishes the goal of reducing block requests in proportion to the fan-out blob requests.

**Comment by kasey** [comment]
note from call with Nishant, treat more than `MAX_REQUEST_BLOCKS_DENEB` context encoded values in the response in `readChunkEncodedBlobs` as an error, similar to the block decoding rpc implementation.

File: beacon-chain/execution/engine_client.go [hunk 22]
```diff
@@ -679,7 +684,11 @@ func (s *Service) ReconstructFullBellatrixBlockBatch(
 	// For blocks that are pre-merge we simply reconstruct them via an empty
 	// execution payload.
 	for _, realIdx := range zeroExecPayloads {
-		payload := buildEmptyExecutionPayload()
+		bblock := blindedBlocks[realIdx]
```
[the remark this answers is not in the document]
Comment (kasey) [reply beacon-chain/execution/engine_client.go:687]: I will rebase deneb integration and then rebase this, so that we can resolve conflicts before merging this pr.

File: beacon-chain/sync/initial-sync/blocks_fetcher_utils.go (hunk 13 above)
Comment (kasey) [reply beacon-chain/sync/initial-sync/blocks_fetcher_utils.go:247]: discussed in live review, will add comments

File: beacon-chain/blockchain/testing/mock.go (hunk 10 above)
Comment (kasey) [reply beacon-chain/blockchain/testing/mock.go:186]: reverting this

File: beacon-chain/sync/initial-sync/blocks_fetcher.go (hunk 19 above)
Comment (kasey) [reply beacon-chain/sync/initial-sync/blocks_fetcher.go:422]: It would throw the alignment off for the next block, so the roots wouldn't match, which would result in `errMismatchedBlobBlockRoot`. Unless it was the last index of the last block in the list, in which case we wouldn't notice it.

File: beacon-chain/sync/initial-sync/blocks_fetcher_utils.go (hunk 13 above)
Comment (kasey) [reply beacon-chain/sync/initial-sync/blocks_fetcher_utils.go:252]: added

File: beacon-chain/sync/initial-sync/blocks_fetcher.go (hunk 18 above)
Comment (kasey) [reply beacon-chain/sync/initial-sync/blocks_fetcher.go]: the blobs are sorted, so if we see a different blob at a given position, that means the blob expected in that position is missing.

File: beacon-chain/sync/initial-sync/round_robin.go (hunk 15 above)
Comment (kasey) [reply beacon-chain/sync/initial-sync/round_robin.go]: as discussed in our call, these changes are downstream of the change to use the `BlockWithVerifiedBlobs` container. When we get deneb-integration merged to develop we can make changes to add ROBlock in more places.

**Review by nisdas — APPROVED** [review]
